group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformAssignor.java

```diff
@@ -0,0 +1,241 @@
+                } else {
+                    partitionRacks = new HashMap<>();
+                    allPartitionRacks = new HashSet<>();
+                    topicIdPartitions.forEach(tp -> {
+                        Set<String> racks = subscribedTopicDescriber.racksForPartition(tp.topicId(), tp.partition());
+                        partitionRacks.put(tp, racks);
+                        if (!racks.isEmpty()) allPartitionRacks.addAll(racks);
+                    });
+                }
+
+                if (useRackAwareAssignment(membersByRack.keySet(), allPartitionRacks, partitionRacks)) {
+                    this.memberRacks = new HashMap<>(assignmentSpec.members().size());
+                    membersByRack.forEach((rack, rackConsumers) -> rackConsumers.forEach(c -> memberRacks.put(c, rack)));
+                    this.partitionRacks = partitionRacks;
+                } else {
+                    this.memberRacks = Collections.emptyMap();
+                    this.partitionRacks = Collections.emptyMap();
+                }
+
+                numMembersByPartition = partitionRacks.entrySet().stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().stream()
+                        .map(r -> membersByRack.getOrDefault(r, Collections.emptyList()).size())
+                        .reduce(0, Integer::sum)));
+            }
+
+            /**
+             * Determines if there's a mismatch between the memberId's rack and the partition's replica racks.
+             *
+             * <p> Mismatch conditions (returns {@code true}):
+             * <ul>
+             *     <li> Consumer lacks an associated rack.</li>
+             *     <li> Partition lacks associated replica racks.</li>
+             *     <li> Consumer's rack isn't among the partition's replica racks.</li>
+             * </ul>
+             *
+             * @param memberId      The memberId identifier.
+             * @param tp            The topic partition in question.
+             * @return {@code true} for a mismatch; {@code false} if member and partition racks exist and align.
+             */
+            protected boolean racksMismatch(String memberId, TopicIdPartition tp) {
+                String consumerRack = memberRacks.get(memberId);
+                Set<String> replicaRacks = partitionRacks.get(tp);
+                return consumerRack == null || (replicaRacks == null || !replicaRacks.contains(consumerRack));
+            }
+
+            /**
+             * Sorts the given list of partitions based on the number of consumers available for each partition
+             * in a rack-aware manner.
+             *
+             * @param partitions    The list of partitions to be sorted.
+             * @return A sorted linked list of partitions with potential members in the same rack.
+             */
+            protected List<TopicIdPartition> sortPartitionsByRackConsumers(List<TopicIdPartition> partitions) {
```

Comment:
I don't think I understand the question

Reply:
numMembersByPartition is a map that's part of the rack info which tracks how many members are in the same rack as the partition. The sum of those members is used to sort them right, that's the "rack aware manner". Sorting happens as expected since I put in print statements to verify them but also without that the assignments wouldn't be as expected and we do a 1:1 check on every assignment in the test cases